### PR TITLE
Feature/api revisited

### PR DESCRIPTION
**Pull request by protich**
Adds email parsing to http api controller - useful for remote piping
Change how api controllers exit - important for local piping 
Api error logging
Other related changes :-1:

### CODE REVIEW

**Comment by protich** [comment]
@greezybacon - Please review the pull request when you get a moment. I did my best to use patches (`git add -p`) to commit related changes in the order of review. 

I still have one pending commit, including DB patch related to API status. 

PS: I lied more like 3-4 more commits :)

File: include/class.ticket.php [hunk 1]
```diff
@@ -1379,40 +1379,43 @@ function release() {
...
+                .', email_mid='.db_input($vars['mid'])
+                .', headers='.db_input($vars['header']);
             db_query($sql);
         }
 
-        if($newticket) return $msgid; //Our work is done...
+        if(!$alerts) return $msgid; //Our work is done...
 
         $autorespond = true;
         if ($autorespond && $headers && TicketFilter::isAutoResponse(Mail_Parse::splitHeaders($headers)))
```
Comment (greezybacon) [inline include/class.ticket.php]: `$headers` => `$vars['header']`

File: include/class.ticket.php [hunk 2]
```diff
@@ -1762,8 +1766,44 @@ function uploadAttachments($files, $refid, $type) {
...
+            //Do error checking...
+            if ($checkFileTypes && !$ost->isFileTypeAllowed($info))
+                $info['error'] = 'Invalid file type (ext) for '.Format::htmlchars($info['name']);
+            elseif ($info['encoding'] && !strcasecmp($info['encoding'], 'base64')) {
+                if(!($info['data'] = base64_decode($info['data'], true)))
+                    $info['error'] = sprintf('%s: Poorly encoded base64 data', Format::htmlchars($info['name']));
+            }
+
+            if($info['error']) {
+                $this->logNote('File Import Error', $error, 'SYSTEM', false);
```
Comment (greezybacon) [inline include/class.ticket.php]: `$error` is undefined. `$info['error']` maybe?

File: include/class.ticket.php [hunk 3]
```diff
@@ -1735,10 +1738,11 @@ function pdfExport($psize='Letter', $notes=false) {
```
Comment (greezybacon) [inline include/class.ticket.php:1737]: It doesn't show up here, but line 1480, `$errors` is passed to postReply completely uninitialized and isn't used in the rest of the function.

Comment (protich) [inline include/class.ticket.php:1737]: Fixed in another pull request - commit 2f1ae23

**Comment by greezybacon** [comment]
Looks pretty good. It passes all the sniff tests.

**Comment by protich** [comment]
I did new test install and an upgrade. You can sign-off and merge.

**Pull request merged**